Sources/ComplexModule/ElementaryFunctions.swift

```diff
@@ -154,78 +154,136 @@ extension Complex /*: ElementaryFunctions */ {
+    // to -1 that 1 + z is computed exactly. If |1+z| => 2, we're far
+    // enough outside the circle that we can simply use log(1+z)--1+z
+    // may round, but that rounding doesn't cause problematic error in
+    // the final result.
+    if 2*r2 <= 1 && !(r2 < 2) { return log(1 + z) }
+    // Here we're in the annulus ½ < |1+z| < 2, so we need to be careful
+    // about cancellation. Instead of computing the real part as
+    // log(|1+z|), we use log(onePlus: 2x + |z|²)/2. This is a cute trick
+    // that Nevin suggested on the swift forums, which works because:
     //
     //     Re(log 1+z) = (log 1+z + log 1+z̅)/2
-    //                 = log((1+z)(1+z̅)/2
+    //                 = log((1+z)(1+z̅))/2
     //                 = log(1+z+z̅+zz̅)/2
     //
-    // This behaves well close to zero, because the z+z̅ term dominates
-    // and is computed exactly. Away from zero, cancellation occurs near
-    // the circle x(x+2) + y^2 = 1, but everywhere along this curve we
-    // have |Im(log 1+z)| >= π/2, so the relative error in the complex
-    // norm is well-controlled. We can take advantage of FMA to further
-    // reduce the cancellation error and recover a good error bound.
-    //
-    // The other common implementation choice for log1p is Kahan's trick:
+    // https://forums.swift.org/t/elementaryfunctions-compliance-for-complex/37903/3
     //
-    //     w := 1+z
-    //     return z/(w-1) * log(w)
+    // So now we just need to compute z+z̅+zz̅ == x(2+x) + y², carefully.
+    // 2+x can round; we have -3 < x < 1, so we can compute a head-tail
+    // sum easily. It's easiest to see why this is exact with a two-case
+    // analysis:
     //
-    // But this actually doesn't do as well as Nevin's approach does,
-    // and requires a complex division, which we want to avoid when we
-    // can do so.
-    var a = 2*z.x
-    // We want to add the larger term first (contra usual guidance for
-    // floating-point error optimization), because we're optimizing for
-    // the catastrophic cancellation case; when that happens adding the
-    // larger term via FMA is always exact. When cancellation doesn't
-    // happen, the simple relative error bound carries through the
-    // rest of the computation.
-    let large = max(z.x.magnitude, z.y.magnitude)
-    let small = min(z.x.magnitude, z.y.magnitude)
-    a.addProduct(large, large)
-    a.addProduct(small, small)
-    // If r2 overflowed, then |z| ≫ 1, and so log(1+z) = log(z).
-    guard a.isFinite else { return log(z) }
-    // Unlike log(z), we do not need to worry about what happens if a
-    // underflows.
-    return Complex(
-      RealType.log(onePlus: a)/2,
-      RealType.atan2(y: z.y, x: 1+z.x)
-    )
+    //  	If x is in [-3, -1], everything is exact because of Sterbenz.
+    //    If x is in (-1, +1], 2 + x will generally round, but 2 > |x|,
+    //      so no sorting is needed to order the tail computation.
+    return log(1+z)
```

Comment:
This doesn’t appear to use the careful calculation that was described in the comments.

Reply:
Yeah, I'm revising it to handle some other cases, but got side-tracked into improving log( ) first.

Reply:
(i.e. reviewing log(onePlus:) is wasted effort, I just hacked a merge together so it would build and I could do some testing on log( ).)